content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,260 @@
+  the node.
+
+* `spec.containers[].resources.limits`
+
+  When you specify the resource limit for containers in a Pod, the kubelet enforces 
+  those limits so that the running containers are not allowed to use more of those 
+  resources than the limits you set.
+
+When a node with enough capacity to serve the requested resources has been identified, 
+the kubelet starts the containers. When the kubelet starts a container as a part of a Pod, 
+kubelet passes the container's requests and limits for CPU and memory to the container 
+runtime. On Linux, the container runtime typically configures kernel cgroups as follows:
+
+* CPU limit
+
+  CPU is considered a compressible resource. Linux kernel periodically checks to see if
+  the CPU limit for a cgroup is exceeded. If your app starts hitting your CPU limits, 
+  Kubernetes starts throttling your container, giving your app potentially worse 
+  performance. However, it won’t be terminated. That is what "compressible" means.
+  
+* CPU request
+
+  The CPU request typically defines a weighting. If several different containers (cgroups)
+  want to run on a contended system, workloads with larger CPU requests are allocated more 
+  CPU time than workloads with small requests.
+  
+* Memory limit
+  
+  The memory limit specifies the memory limit, beyond which if the container tries to allocate 
+  more memory, Linux kernel will terminate it with an OOM (Out of Memory) kill.
+
+  In cgroup v1, `memory.limit_in_bytes` interface is used to set the memory usage limit.
+  However, unlike CPU memory can't be compressed as cgroup v1 doesn't provide any interface
+  to specify memory usage throttle limit. 
+
+  In cgroup v2, `memory.max` is analogous to `memory.limit_in_bytes` in cgroupv1.
+  And `memory.high` interface is introduced to specify memory usage throttling
+  limit. Memory QoS uses `memory.max` to specify memory limits, and `memory.high` 
+  to throttle workload approaching its memory limit, ensuring that the system is
+  not overwhelmed by OOM kills from containers trying to use memory beyond specified
+  limits.
+
+* Memory request
+
+  The memory request is mainly used by kube-scheduler during (Kubernetes) Pod 
+  scheduling. In cgroup v1, there are no controls to specify the minimum amount
+  of memory the cgroup must always retain. Hence, the container runtime did not
+  use the value of requested memory set in the Pod Spec. 
+
+  cgroup v2 introduced `memory.min` interface to specify the minimum amount of memory
+  a cgroup must always retain. If the memory usage of a cgroup is within its effective
+  min boundary, the cgroup’s memory won’t be reclaimed under any conditions. If 
+  there's no unprotected reclaimable memory available, OOM killer is invoked. Memory
+  QoS maps `memory.min` to `spec.containers[].resources.requests.memory` to ensure
+  the availability of memory for containers in Kubernetes Pods. 
```

Comment:
maybe `to ensure the availability of memory for containers in Kubernetes Pods - something that was not possible before cgroup v2 and MemoryQoS?`